hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java

```diff
@@ -0,0 +1,160 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import com.addthis.basis.util.Parameter;
+
+import com.addthis.bundle.core.Bundle;
+import com.addthis.hydra.task.source.TaskDataSource;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class MapFeederForkJoin implements Runnable {
+    private static final Logger log = LoggerFactory.getLogger(MapFeederForkJoin.class);
+
+    private static final int QUEUE_DEPTH = Parameter.intValue("task.queue.depth", 100);
+
+    // state control
+    private final AtomicBoolean errored = new AtomicBoolean(false);
+    private boolean hasClosedStreams = false; // not shared with MapperTasks
+
+    // enclosing task
+    private final StreamMapper task;
+    private final TaskDataSource source;
+
+    // mapper task controls
+    private final ForkJoinPool mapperPool;
+    private final int parallelism;
+    private final Semaphore enqueuePermits;
+
+    private long start = System.currentTimeMillis();
+    private long totalBundles = 0;
+
+    public MapFeederForkJoin(StreamMapper task, TaskDataSource source, int parallelism) {
+        this.source = source;
+        this.task = task;
+        this.parallelism = parallelism;
+        this.mapperPool = new ForkJoinPool(parallelism);
+        this.enqueuePermits = new Semaphore(QUEUE_DEPTH);
+    }
+
+    @Override public void run() {
+        log.info("starting {} thread(s) for src={}", parallelism, source);
+        try {
+            if (source.isEnabled()) {
+                while (fillBuffer()) {
+                    if (Thread.interrupted()) {
+                        closeSourceIfNeeded();
+                    }
+                }
+            }
+            closeSourceIfNeeded();
+            joinProcessors();
+            log.info("all ({}) task threads exited; sending taskComplete", parallelism);
+            logBundleThroughput();
+            // run in different threads to isolate them from interrupts. ie. "taskCompleteUninterruptibly"
+            // join awaits completion, is uninterruptible, and will propagate any exception
+            CompletableFuture.runAsync(task::taskComplete).join();
+        } catch (Throwable t) {
+            logBundleThroughput();
+            handleUncaughtThrowable(t);
+        }
+        log.debug("task feeder exited");
+    }
+
+    private void logBundleThroughput() {
+        long elapse = (System.currentTimeMillis() - start) / 1000;
+        log.info("{} bundles processed in {} seconds (avg rate={}/s)", totalBundles, elapse, totalBundles / elapse);
+    }
+
+    /**
+     * Immediately halting is the best way we have of ensuring errors are reported without a significantly
+     * more involved system.
+     */
+    private void handleUncaughtThrowable(Throwable t) {
+        if (errored.compareAndSet(false, true)) {
+            log.error("unrecoverable error in task feeder or one of its mapper threads. immediately halting jvm", t);
+            Runtime.getRuntime().halt(1);
```

Comment:
it's in the latest snapshot, and I mean specifically the logic contained in the handler (the Runtime.getRuntime().halt(1) call). The current namespacing is `Shutdown.exit(int)`